include/oneapi/dpl/internal/random_impl/counter_based_engines_stuff.h

```diff
@@ -0,0 +1,91 @@
+// -*- C++ -*-
+//===-- philox-engine.h ------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+//
+// Abstract:
+//
+// Header file provides stuff functionality for counter-based RNG engines.
+//
+//   fffmask<Uint, w> - the Uint with the low w bits set
+
+
+#ifndef _ONEDPL_CTR_ENGINES_STUFF_H
+#define _ONEDPL_CTR_ENGINES_STUFF_H
+
+#include <concepts>
+#include <iterator>
+#include <utility>
+#include <tuple>
+#include <limits>
+
+namespace detail {
+template <typename U, unsigned W,
+          typename = ::std::enable_if_t<std::is_unsigned_v<U> && (W <= std::numeric_limits<U>::digits)>>
+constexpr U fffmask = W ? (U(~(U(0))) >> (std::numeric_limits<U>::digits - W)) : 0;
+
+// For unpacking variadic of constants into two arrays:
+template <typename UIntType, typename Tuple, size_t... Is>
+constexpr auto get_even_array_from_tuple(Tuple t, std::index_sequence<Is...>) {
+    return std::array<UIntType, std::index_sequence<Is...>::size()>{ std::get<Is * 2>(t)... };
+}
+
+template <typename UIntType, typename Tuple, size_t... Is>
+constexpr auto get_odd_array_from_tuple(Tuple t, std::index_sequence<Is...>) {
+    return std::array<UIntType, std::index_sequence<Is...>::size()>{ std::get<Is * 2 + 1>(t)... };
+}
+
+// Implement w-bit mulhilo with an 2w-wide integer - returns
+// the w hi and w low bits of the 2w-bit product of a and b.
+template <typename UIntType, unsigned W>
+static std::pair<UIntType, UIntType> mulhilo(UIntType a, UIntType b)
+{
+    using result_type = UIntType;
+
+    result_type res_hi, res_lo;
+    /* multiplication fits standard types */
+    if(W <= 32) {
+        uint_fast64_t mult_result = (uint_fast64_t)a * (uint_fast64_t)b;
+        res_hi = mult_result >> W;
+        res_lo =  mult_result & detail::fffmask<result_type, W>;
+    }
+    /* pen-pencil multiplication by 32-bit chunks */
+    else if(W == 64) {
+        res_lo = a * b;
```

Comment:
I assume this approach should also work for W < 64, you do it as for 64 and then apply & mask

or as an else here we can check if compiler supports __int128 (e.g. if CLANG or GCC) and have a multiply as 2 128-bit numbers

Reply:
> I assume this approach should also work for W < 64, you do it as for 64 and then apply & mask

Yep, I've corrected the condition

> or as an else here we can check if compiler supports __int128 (e.g. if CLANG or GCC) and have a multiply as 2 128-bit numbers
Decided with @paveldyakov to implement this later, as an improvement of the current version. The condition of `__int128` type enabling should be defined very carefully to prevent compilation errors on user's size.